pkg/controller/certificates/issuing/secret_manager.go

```diff
@@ -0,0 +1,222 @@
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	"github.com/jetstack/cert-manager/pkg/controller/certificates/internal/secretsmanager"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+)
+
+// ensureSecretData ensures that the Certificate's Secret is up to date with
+// non-issuing condition related data. Currently only reconciles on Annotations
+// and Labels from the Certificate's SecretTemplate.
+func (c *controller) ensureSecretData(ctx context.Context, log logr.Logger, crt *cmapi.Certificate) error {
+	dbg := log.V(logf.DebugLevel)
+
+	// Retrieve the Secret which is associated with this Certificate.
+	secret, err := c.secretLister.Secrets(crt.Namespace).Get(crt.Spec.SecretName)
+
+	// Secret doesn't exist so we can't do anything. The Certificate will be
+	// marked for a re-issuance and the resulting Secret will be evaluated again.
+	if apierrors.IsNotFound(err) {
+		dbg.Info("secret not found", "error", err.Error())
+		return nil
+	}
+
+	// This error is transient, return error to be retried on the rate limiting
+	// queue.
+	if err != nil {
+		return err
+	}
+
+	secret = secret.DeepCopy()
+
+	var data secretsmanager.SecretData
+	if secret.Data != nil {
+		data = secretsmanager.SecretData{
+			PrivateKey:  secret.Data[corev1.TLSPrivateKeyKey],
+			Certificate: secret.Data[corev1.TLSCertKey],
+			CA:          secret.Data[cmmeta.TLSCAKey],
+		}
+	}
+
+	log = log.WithValues("secret", secret.Name)
+
+	// Check whether the Certificate's SecretTemplate matches that on the Secret.
+	secretTemplateMatchManagedFields, err := c.secretTemplateMatchesManagedFields(crt, secret, data)
+	if err != nil {
+		// An error here indicates that the managed fields are malformed and the
+		// decoder doesn't understand the managed fields on the Secret, or the
+		// signed certificate data could not be decoded. There is nothing more the
+		// controller can do here, so we exit nil so this controller doesn't end in
+		// an infinite loop.
+		log.Error(err, "failed to determine whether the SecretTemplate matches Secret")
+		return nil
+	}
+
+	// - secretTemplateMatchesSecret: If a key or value changed on the
+	// Annotations or Labels in the SecretTemplate, the SecretTemplate will not
+	// match the Annotations or Labels on the Secret.
+	// - secretTemplateMatchManagedFields: If a key was removed on the
+	// SecretTemplate, then the managed fields on the Secret won't match.
+	// In either case, the Secret needs to be re-reconciled with the Secrets
+	// Manager.
+	if !secretTemplateMatchesSecret(crt, secret) || !secretTemplateMatchManagedFields {
+		log.Info("mismatch between SecretTemplate and Secret, updating Secret annotations/labels")
```

Comment:
Do we have a documented or undocumented strategy for when to use different log levels?

Reply:
Not a perfect solution, but any time I want to know "which level should I use??", I "guess" from looking at the other uses in the codebase:

```sh
rg 'log.*(Info|Error)'
```

or 

```sh
grep -E 'log.*(Info|Error)' -R --include=\*.go
```

We could use this data to add more information to https://cert-manager.io/docs/contributing/coding-conventions/.

Reply:
I don't think we have these things specifically written down anywhere, but these are the constant names which give a hint https://github.com/jetstack/cert-manager/blob/master/pkg/logs/logs.go#L44-L49.

One could argue that this log line could be reduced to Debug.

Reply:
Here is the usage statistics:

```sh
$ rg 'log[^\s]*(Info|Error)' --only-matching --no-filename --no-heading --no-line-number | sort | uniq -c | sort -n --reverse
    188 log.Error
    152 log.V(logf.DebugLevel).Info
     44 log.V(logf.InfoLevel).Info
     20 logf.V(logf.DebugLevel).Info
     19 log.Info
     16 log.V(logf.TraceLevel).Info
     12 log.V(logf.WarnLevel).Info
     10 logf.V(logf.WarnLevel).Info
      9 logf.V(logf.InfoLevel).Info
      4 log.V(logf.ErrorLevel).Info
      4 logf.Log.V(logf.DebugLevel).Info
      3 logf.Log.V(logf.InfoLevel).Info
      3 logf.Log.Error
      2 log.V(logs.DebugLevel).Info
      2 log.V(logf.Info
      2 log.V(logf.ExtendedInfo
      2 log.Logf("Error
      2 logf.Log.Info
      2 logf.DebugLevel).Info
      1 log.V(2).Info
      1 logf.WithInfof(log.V(logf.DebugLevel)).Info
      1 logf.WarnLevel).Info
      1 logf.Log.V(logf.WarnLevel).Info
      1 logf.FromContext(ctx).V(logf.DebugLevel).Info
```

- Wow, we use log.Error a lot. 
- I see two `log.V(2).Info` which should be logf.InfoLevel I think

Reply:
I think you both picked up on what I was alluding to: a clear guide on what to log at which level and when.

The constants links (https://github.com/jetstack/cert-manager/blob/master/pkg/logs/logs.go#L42) to https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md which is a step in that direction.

I think we should have a clear definition in our docs, or simply link to someone else's clear definition.

I've had many many long discussions on the philosophy of logging, and usually there are wildly varying opinions on when, how and what. I think the important thing is that there is a documented definition of when and how, and what the purpose of a log statement is (e.g. give users clear information on how to proceed if something goes wrong or is unrecoverable)

I'll bring it up in a separate forum.